core/src/main/java/io/atomix/core/multimap/impl/CachingAsyncConsistentMultimap.java

```diff
@@ -0,0 +1,180 @@
+   * @param cacheSize  the maximum size of the cache
+   */
+  public CachingAsyncConsistentMultimap(AsyncConsistentMultimap<K, V> backingMap, int cacheSize) {
+    super(backingMap);
+    cache = CacheBuilder.newBuilder()
+        .maximumSize(cacheSize)
+        .build(CacheLoader.from(CachingAsyncConsistentMultimap.super::get));
+    cacheUpdater = event -> {
+      V oldValue = event.oldValue();
+      V newValue = event.newValue();
+      CompletableFuture<Versioned<Collection<? extends V>>> future = cache.getUnchecked(event.key());
+      switch (event.type()) {
+        case INSERT:
+          if (future.isDone()) {
+            Versioned<Collection<? extends V>> oldVersioned = future.join();
+            Versioned<Collection<? extends V>> newVersioned = new Versioned<>(
+                ImmutableSet.<V>builder().addAll(oldVersioned.value()).add(newValue).build(),
+                oldVersioned.version(),
+                oldVersioned.creationTime());
+            cache.put(event.key(), CompletableFuture.completedFuture(newVersioned));
+          } else {
+            cache.put(event.key(), future.thenApply(versioned -> new Versioned<>(
+                ImmutableSet.<V>builder().addAll(versioned.value()).add(newValue).build(),
+                versioned.version(),
+                versioned.creationTime())));
+          }
+          break;
+        case REMOVE:
+          if (future.isDone()) {
+            Versioned<Collection<? extends V>> oldVersioned = future.join();
+            cache.put(event.key(), CompletableFuture.completedFuture(new Versioned<>(oldVersioned.value()
+                .stream()
+                .filter(value -> !Objects.equals(value, oldValue))
+                .collect(Collectors.toSet()), oldVersioned.version(), oldVersioned.creationTime())));
+          } else {
+            cache.put(event.key(), future.thenApply(versioned -> new Versioned<>(versioned.value()
+                .stream()
+                .filter(value -> !Objects.equals(value, oldValue))
+                .collect(Collectors.toSet()), versioned.version(), versioned.creationTime())));
+          }
+          break;
+        default:
+          break;
+      }
+    };
+    statusListener = status -> {
+      log.debug("{} status changed to {}", this.name(), status);
+      // If the status of the underlying map is SUSPENDED or INACTIVE
+      // we can no longer guarantee that the cache will be in sync.
+      if (status == SUSPENDED || status == INACTIVE) {
+        cache.invalidateAll();
+      }
+    };
+    super.addListener(cacheUpdater);
+    super.addStatusChangeListener(statusListener);
+  }
+
+  @Override
+  public CompletableFuture<Boolean> containsKey(K key) {
+    return get(key).thenApply(value -> value != null && !value.value().isEmpty());
+  }
+
+  @Override
+  public CompletableFuture<Boolean> put(K key, V value) {
+    return super.put(key, value)
+        .whenComplete((r, e) -> cache.invalidate(key));
```

Comment:
invalidate on put, or update?

Reply:
Both. The cache can have a future that returns `null` from a prior read on an empty key.